### ISSUE TITLE
minio first letter capitalization of metadata keys with boto3 sdk

### ISSUE THREAD

**Issue by eedwards-sk** [minio/minio#6471]
## Expected Behavior
`object.metadata` contains key `sha256`

## Current Behavior
`object.metadata` contains key `Sha256`

## Possible Solution
I suspect maybe minio is auto upcasing the header names when returning them with boto3?
e.g. the header from this example on an actual aws bucket viewed from aws console is `x-amz-meta-sha256`
I suspect maybe minio returns something like `X-Amz-Meta-Sha256` or similar, where it auto-upcases the first letter?

## Steps to Reproduce (for bugs)
given a local minio environment at `http://minio.service.local.consul:9000` with region `local`
and credentials stored in a profile `minio-local`
and a bucket named `local-pki`
and a file named `root-ca.pem`

1. upload the file to minio, using the aws cli, and specifying metadata with a key of `SHA256`:
```
AWS_PROFILE=minio-local AWS_DEFAULT_REGION=local aws --endpoint http://minio.service.local.consul:9000/ s3 cp --metadata SHA256=c3f947b8a5a85db8961554d0ecdfef0c944ac00b55fec5ea7f5d12fdde7c5bd5 root-ca.pem s3://local-pki/root-ca.pem
```

2. view the metadata using the aws cli, note it's lowercased
```
AWS_PROFILE=minio-local AWS_DEFAULT_REGION=local aws --endpoint http://minio.service.local.consul:9000/ s3api head-object --bucket local-pki --key root-ca.pem
```

3. configure boto3 and view the metadata, note it's upper cased the first letter
```
Python 3.6.6 (default, Sep 11 2018, 15:24:10)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import boto3
>>> session = boto3.session.Session(region_name='local', profile_name='minio-local')
>>> s3 = session.resource('s3', endpoint_url='http://minio.service.local.consul:9000', use_ssl=False)
>>> object = s3.Object('local-pki', 'root-ca.pem')
>>> object.metadata
{'Sha256': 'c3f947b8a5a85db8961554d0ecdfef0c944ac00b55fec5ea7f5d12fdde7c5bd5'}
```

an actual aws bucket (not minio) returns `sha256` for the key name

## Context
trying to access a metadata key in python, where it gets mapped to a dict and becomes case sensitive... when using minio, it comes back with first letter capitalized, unlike aws

## Regression
N\A

## Your Environment
VERSION
2018-08-02T23:11:36Z
MEMORY
Used: 5.4 MB | Allocated: 37 MB | Used-Heap: 5.4 MB | Allocated-Heap: 9.7 MB
PLATFORM
Host: 84372e135c56 | OS: linux | Arch: amd64
RUNTIME
Version: go1.10.2 | CPUs: 4

running inside docker on mac osx

**Comment by eedwards-sk** [comment]
is minio properly lowercasing headers?

According to https://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html
under **Constructing the CanonicalizedAmzHeaders Element**

the first step states

>Convert each HTTP header name to lowercase. For example, 'X-Amz-Date' becomes 'x-amz-date'.

also see https://forums.aws.amazon.com/thread.jspa?threadID=32168

**Comment by harshavardhana** [comment]
Canonization is only for signature calculation http headers are case insensitive

**Comment by eedwards-sk** [comment]
Not sure what that means exactly or the implications?

As it stands, I can't use minio in place of s3 because it does not lower case the metadata key names the same way that aws s3 does.

According to the link I referenced, it's likely related to how headers are generated or handled, since that's how metadata is stored.

**Comment by harshavardhana** [comment]
> Not sure what that means exactly or the implications?

There are no implications it is not possible to preserve the case of HTTP headers in transit since this is not mandated by the HTTP protocol 

https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

> HTTP header fields, which include general-header (section 4.5), request-header (section 5.3), response-header (section 6.2), and entity-header (section 7.1) fields, follow the same generic format as that given in Section 3.1 of RFC 822 [9]. Each header field consists of a name followed by a colon (":") and the field value. **Field names are case-insensitive**. The field value MAY be preceded by any amount of LWS, though a single SP is preferred. Header fields can be extended over multiple lines by preceding each extra line with at least one SP or HT. Applications ought to follow "common form", where one is known or indicated, when generating HTTP constructs, since there might exist some implementations that fail to accept anything

So there is no way for the server to know what was the case for HTTP headers were sent in the first place. So having 'Sha256' is same as 'sha256', we rely on Go HTTP server which automatically does this before we receive the I/O. 

> According to the link I referenced, it's likely related to how headers are generated or handled, since that's how metadata is stored.

What you mentioned is for calculating signature for authentication which we already do, it has nothing to do with metadata HTTP headers. A canonical representation of `x-amz-meta-sha256` in HTTP terms is `X-Amz-Meta-Sha256` so for a server both headers are same. This complies with the RFC and any clients which require casesentivity means they are intentionally using it incorrectly trying to rely on HTTP headers to be case sensitive. 

Testing a scenario like this with S3 one can observe that they treat the headers case insensitively, i.e they don't treat 'Sha256' different than 'sha256'
```
$ curl -v -H "X-Amz-Meta-Sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3" -H "X-Amz-Meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3" -H "x-amz-meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3" -H "X-amz-Meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3" --upload-file /etc/issue https://s3.amazonaws.com/vadmeste/prefix/meta

$ curl -i https://s3.amazonaws.com/vadmeste/prefix/meta
HTTP/1.1 200 OK
x-amz-id-2: PDrTPr1jqa915ct8WteuGMyA5Sf/7tsab0ZSIZaIFM7JWhj3fju1cRzqgxpY0QR5DC9km5p6N0M=
x-amz-request-id: A2CD149C4C89B730
Date: Sun, 16 Sep 2018 21:26:08 GMT
Last-Modified: Sun, 16 Sep 2018 21:25:40 GMT
ETag: "38103f9a76bb14e2abbc051b65722cff"
x-amz-meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3,8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3,8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3,8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3
x-amz-version-id: fk9ln8tfFKd55Ex9.mKY_NdJY13XqGfb
Accept-Ranges: bytes
Content-Type: binary/octet-stream
Content-Length: 26
Server: AmazonS3

Ubuntu 18.04.1 LTS \n \l
```

This is exactly how Minio behaves
```
$ curl -v -H "X-Amz-Meta-Sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3" -H "X-Amz-Meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3" -H "x-amz-meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3" -H "X-amz-Meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3" --upload-file /etc/issue https://play.minio.io:9000/vadmeste/prefix/meta
*   Trying 147.75.201.93...
* TCP_NODELAY set
* Connected to play.minio.io (147.75.201.93) port 9000 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-CHACHA20-POLY1305
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=play.minio.io
*  start date: Aug 21 07:20:41 2018 GMT
*  expire date: Nov 19 07:20:41 2018 GMT
*  subjectAltName: host "play.minio.io" matched cert's "play.minio.io"
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
> PUT /vadmeste/prefix/meta HTTP/1.1
> Host: play.minio.io:9000
> User-Agent: curl/7.58.0
> Accept: */*
> X-Amz-Meta-Sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3
> X-Amz-Meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3
> x-amz-meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3
> X-amz-Meta-sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3
> Content-Length: 26
> Expect: 100-continue
> 
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 200 OK
< Accept-Ranges: bytes
< Content-Security-Policy: block-all-mixed-content
< Etag: "38103f9a76bb14e2abbc051b65722cff"
< Server: Minio/DEVELOPMENT.2018-09-13T21-51-05Z (linux; amd64)
< Vary: Origin
< X-Amz-Request-Id: 1554FE88ECC0A12B
< X-Xss-Protection: 1; mode=block
< Date: Sun, 16 Sep 2018 21:27:16 GMT
< Content-Length: 0
< 
* Connection #0 to host play.minio.io left intact

$ curl -i https://play.minio.io:9000/vadmeste/prefix/meta
HTTP/1.1 200 OK
Accept-Ranges: bytes
Content-Length: 26
Content-Security-Policy: block-all-mixed-content
Etag: "38103f9a76bb14e2abbc051b65722cff"
Last-Modified: Sun, 16 Sep 2018 21:27:16 GMT
Server: Minio/DEVELOPMENT.2018-09-13T21-51-05Z (linux; amd64)
Vary: Origin
X-Amz-Meta-Sha256: 8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3,8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3,8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3,8ba8085ad1bb937158eb6db5de263219ff69a7ca87637425e85f3fcf683cd1a3
X-Amz-Request-Id: 1554FE8F17FD9015
X-Xss-Protection: 1; mode=block
Date: Sun, 16 Sep 2018 21:27:43 GMT
Content-Type: text/plain; charset=utf-8

Ubuntu 18.04.1 LTS \n \l
```

The only difference is canonical version of the HTTP header which is as per RFC valid and clients should handle it correctly and not rely on case sensitivity.  Moved the label for this issue as working as intented, this is simply a bug on the client not server.

**Comment by eedwards-sk** [comment]
Hey thanks! That shows a very detailed repro of exactly what I'm taking about.

`x-amz-meta-sha256` vs `X-Amz-Meta-Sha256`

If it's not possible to preserve it in transit, then i'm curious why it's clearly and repeatably different between aws and minio? LOTS of people use the boto3 sdk and metadata, but I couldn't find any evidence of this being an issue for anyone using official aws s3 buckets.

I could interpret 'working as intended' meaning it's still being RFC-soandso **compliant**, but there's clearly still a **compatibility** issue here (and my understanding is that minio should be aws compatible).

So I'm hoping you're saying it's still a valid bug and not a 'won't fix'? I'd love to be able to keep using minio, it's very helpful so far!

**Comment by harshavardhana** [comment]
It's not a valid bug as I explained yes there are differences in the HTTP headers but that is allowed as per RFC so a server can happily send HTTP headers in any case whatsoever. 

Boto3 trying to keep a case sensitive behavior is wrong that is why client is buggy here. 

```
object.metadata
```

Should simply be populated by lowecasing the headers, if that is expected by applications.

**Comment by eedwards-sk** [comment]
>It's not a valid bug as I explained yes there are differences in the HTTP headers but that is allowed as per RFC so a server can happily send HTTP headers in any case whatsoever.

but that's not what aws does... AWS lowercases the headers, as we saw. So how it that not a bug?

I agree that boto should lowercase the headers if it is trying to be **RFC compatible**, but it's not, because aws always sends lowercased headers.

So isn't it a bug that minio is not compatible? Maybe this is a semantics thing and you don't call broken compatibility bugs.

**Comment by harshavardhana** [comment]
AWS behavior is correct it is allowed in HTTP RFC, likewise we don't do that because Go chooses to canonicalize it which is also correct.

So the bug is in the client which is expected to honor RFC.

**Comment by eedwards-sk** [comment]
Unfortunately I don't believe boto3 will change their behavior to fit other non-aws providers (e.g. minio), so imo the responsibility should be on minio to be compatible.

I'll try opening an issue with boto3 none the less.

(Edit: just to reinforce my stance, they close issues that are due to minio: https://github.com/boto/boto3/issues/1425)

>You should file a bug with minio to tell them they are not accurately replicating the S3 behavior.

Edit2: I've opened https://github.com/boto/boto3/issues/1709 and I leveraged your example, thanks :P

**Comment by eedwards-sk** [comment]
https://github.com/boto/boto3/issues/1709

boto3 has confirmed they will not be changing this

Thus, the important question that needs to be answered:

is minio supposed to be s3 compatible or not?

**If yes: this is a bug**

RFC compliance is not relevant to this issue, as **being RFC compliant does not mean it's compatible with AWS**.

**Comment by harshavardhana** [comment]
> RFC compliance is not relevant to this issue, as **being RFC compliant does not mean it's compatible with AWS**.

RFC compliance is relevant because underneath they are all HTTP clients and AWS S3 is an HTTP based protocol. If they rely on HTTP they better comply with RFC if not then its a broken implementation with specialized behavior which is not HTTP. 

So Minio won't fix this - HTTP RFC compliance wins here. 

> is minio supposed to be s3 compatible or not?

This doesn't invalidate that Minio is not an S3 complaint server, for example, aws-sdk-go, aws-sdk-java all of them handle this appropriately, it just means that boto3 is not willing to honor the HTTP standard.

**Comment by eedwards-sk** [comment]
Except if I have to use specific clients to use aws, then minio is not useful to me, as the entire point of using minio is as a drop in replacement, not to have to deal with a bunch of workarounds.

At that point, minio just becomes another 3rd party object storage solution.

If aws compatibility is actually what minio is for, then you would mirror their behavior whether it fits your ideal of RFC compliance or not.

Is that not clear? I'm just trying to use the tool, not get into semantics, but arguing that "well it's not per spec, aws is wrong" on a tool that is designed to be **aws compatible** makes no sense to me.

(In other words, if you just downcased the meta headers like aws does, then all the tools would work, because that's **what aws does**. Whether they should be or shouldn't be isn't what I opened this issue for)

**Comment by harshavardhana** [comment]
> Is that not clear? I'm just trying to use the tool, not get into semantics, but arguing that "well it's not per spec, aws is wrong" on a tool that is designed to be **aws compatible** makes no sense to me.

As i suggested examples already aws-sdk-go (implicitly because Minio is using the same behavior), aws-sdk-java handle it I am not sure why boto3 is special here. 

https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/ObjectMetadata.html#getUserMetadata--

> public Map<String,String> getUserMetadata()
> Gets the custom user-metadata for the associated object.
> 
> Amazon S3 can store additional metadata on objects by internally representing it as HTTP headers prefixed with "x-amz-meta-". Use user-metadata to store arbitrary metadata alongside their data in Amazon S3. When setting user metadata, callers should not include the internal "x-amz-meta-" prefix; this library will handle that for them. Likewise, when callers retrieve custom user-metadata, they will not see the "x-amz-meta-" header prefix.
> 
> **User-metadata keys are case insensitive and will be returned as lowercase strings, even if they were originally specified with uppercase strings.**
> 
> Note that user-metadata for an object is limited by the HTTP request header limit. All HTTP headers included in a request (including user metadata headers and other standard HTTP headers) must be less than 8KB.
> 
> Returns:
> The custom user metadata for the associated object.
> See Also:
> setUserMetadata(Map), addUserMetadata(String, String)
>

**Comment by harshavardhana** [comment]
Going to close this issue, for now, let's re-open if boto3 closes their issue. A simple JAVA example using aws-sdk-java shows that this is a boto3 issue, so clearly boto3 is ignoring a key requirement. 

```java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.util.UUID;

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.auth.profile.ProfileCredentialsProvider;
import com.amazonaws.regions.Region;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3Client;
import com.amazonaws.services.s3.S3ClientOptions;
import com.amazonaws.services.s3.model.Bucket;
import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.s3.model.ListObjectsRequest;
import com.amazonaws.services.s3.model.ObjectListing;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.S3ObjectSummary;
import com.amazonaws.services.s3.model.ObjectMetadata;

public class S3Sample {
    public static void main(String[] args) throws IOException {
        AWSCredentials credentials = null;
        try {
            credentials = new ProfileCredentialsProvider().getCredentials();
        } catch (Exception e) {
            throw new AmazonClientException(
                    "Cannot load the credentials from the credential profiles file. " +
                    "Please make sure that your credentials file is at the correct " +
                    "location (~/.aws/credentials), and is in valid format.", e);
        }

        AmazonS3 s3 = new AmazonS3Client(credentials);
        Region usEast1 = Region.getRegion(Regions.US_EAST_1);
        s3.setRegion(usEast1);
        s3.setEndpoint("http://localhost:9000");
        S3ClientOptions clientOptions = new S3ClientOptions();
        clientOptions.withPathStyleAccess(true);
        s3.setS3ClientOptions(clientOptions);

        String bucketName = "testbucket";
        String key = "MyObjectKey";

        try {
            System.out.println("Uploading a new object to S3 from a file");
            ObjectMetadata metadata = new ObjectMetadata();
            metadata.addUserMetadata("sha256", "hello");
            s3.putObject(new PutObjectRequest(bucketName, key, createSampleFile()).withMetadata(metadata));

            // Get metadata
            ObjectMetadata rmetadata = s3.getObjectMetadata(bucketName, key);
            System.out.println("X-Amz-Meta-Sha256 " + rmetadata.getUserMetaDataOf("sha256"));
        } catch (AmazonServiceException ase) {
            System.out.println("Caught an AmazonServiceException, which means your request made it "
                    + "to Amazon S3, but was rejected with an error response for some reason.");
            System.out.println("Error Message:    " + ase.getMessage());
            System.out.println("HTTP Status Code: " + ase.getStatusCode());
            System.out.println("AWS Error Code:   " + ase.getErrorCode());
            System.out.println("Error Type:       " + ase.getErrorType());
            System.out.println("Request ID:       " + ase.getRequestId());
        } catch (AmazonClientException ace) {
            System.out.println("Caught an AmazonClientException, which means the client encountered "
                    + "a serious internal problem while trying to communicate with S3, "
                    + "such as not being able to access the network.");
            System.out.println("Error Message: " + ace.getMessage());
        }
    }

    private static File createSampleFile() throws IOException {
      File file = File.createTempFile("aws-java-sdk-", ".txt");
      file.deleteOnExit();
      Writer writer = new OutputStreamWriter(new FileOutputStream(file));
      writer.write("abcdefghijklmnopqrstuvwxyz\n");
      writer.write("01234567890112345678901234\n");
      writer.write("!@#$%^&*()-=[]{};':',.<>/?\n");
      writer.write("01234567890112345678901234\n");
      writer.write("abcdefghijklmnopqrstuvwxyz\n");
      writer.close();

      return file;
    }
}
```

```
$ ant
Buildfile: /home/harsha/aws-sdk/java/aws-java-sdk-1.11.21/samples/AmazonS3/build.xml

run:
     [java] Uploading a new object to S3 from a file
     [java] X-Amz-Meta-Sha256 hello

BUILD SUCCESSFUL
Total time: 2 seconds
```
```
mc stat myminio/testbucket/MyObject
Name      : MyObjectKey
Date      : 2018-09-19 02:54:15 PDT 
Size      : 135B   
ETag      : 24156eb0ff2ed1056106fb2fc4d0ed85 
Type      : file 
Metadata  :
  Content-Type     : text/plain 
  X-Amz-Meta-Sha256: hello 
```

**Comment by eedwards-sk** [comment]
as a follow up, I implemented a workaround which has allowed me to resume using minio, in spite of boto3's bass ackwards implementation of the metadata keys

```
# =============================================================================
# _get_s3_object_checksum
# =============================================================================
def _get_s3_object_checksum(
    s3_object: boto3.resources.base.ServiceResource
) -> boto3.resources.base.ServiceResource:
    # workaround for https://github.com/boto/boto3/issues/1709
    # which results in case-sensitive keys.
    # since it's impossible to end up with two different keys
    # in the metadata dict with the same case-insensitive name
    # (e.g. 'Foo' and 'foo'), then we can assume the lowercased
    # version of the key is the same as the lowercased
    # version of the expected key
    # in addition, when setting the value, the case of the
    # metadata key does not matter, by design

    # loop through each key in the metadata dict
    for key in s3_object.metadata.keys():
        # if the lowercased version of the key
        # matches the lowercased version of the expected key
        if key.lower() == CHECKSUM_METADATA_KEY_NAME.lower():
            # return the actual key's value
            return s3_object.metadata[key]
    # otherwise, if we didn't return, throw a key error
    raise KeyError(f"metadata key '{CHECKSUM_METADATA_KEY_NAME}' not found")
```

While not ideal, and it doesn't solve anything for projects that I do not write, it at least unblocked me for this one use case.

**Comment by vsoch** [comment]
We are also hitting this issue: https://github.com/singularityhub/sregistry-cli/pull/169#issuecomment-451903100 For a workaround I will check for both lower and capitalized metadata keys.

**Comment by harshavardhana** [comment]
> We are also hitting this issue: [singularityhub/sregistry-cli#169 (comment)](https://github.com/singularityhub/sregistry-cli/pull/169#issuecomment-451903100) For a workaround I will check for both lower and capitalized metadata keys.

This has to be fixed upstream @vsoch

**Comment by vsoch** [comment]
Yes, understood. The fix is this one? https://github.com/aws/aws-sdk-net/pull/1143

**Comment by harshavardhana** [comment]
> Yes, understood. The fix is this one? [aws/aws-sdk-net#1143](https://github.com/aws/aws-sdk-net/pull/1143)

Not that's for dotnet instead this is the issue https://github.com/boto/boto3/issues/1709

**Comment by vsoch** [comment]
Excellent thank you.
